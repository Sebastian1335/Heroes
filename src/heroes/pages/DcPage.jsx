import { HeroList } from "../components"

const publisher = 'DC Comics'
export const DcPage = () => {
  return (
    <>
      <h1>DC Comics</h1>
      <hr />
      <HeroList publisher={publisher}/>
    </>
  )
}
