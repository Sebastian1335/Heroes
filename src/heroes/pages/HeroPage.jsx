import { Navigate, useNavigate, useParams } from "react-router-dom"
import {getHeroById} from '../helpers'
import { useMemo } from "react"

export const HeroPage = () => {

  const {heroId} = useParams()//! useParams para recojer la informacion de la URL como variables
  const navigate = useNavigate()

  const hero =useMemo(() => getHeroById(heroId), [heroId]) //! Si al Id no cambia no se vuelve a disparar la funcion, de lo contrario si

  const onNavigateback = () => {
    if (hero.publisher.includes('Marvel')) {
      navigate('/marvel')
    }else{
      navigate('/dc')
    }
  }

  if (!hero) {
    return <Navigate to={"/marvel"}/> //! Es para que te saque a determinada direccion cuando no ingresas un URl valido
  }
  return (
    <div className="row mt-5 animate__animated animate__fadeInLeft">
      <div className="col-4">
        <img 
          src= {`/assets/heroes/${heroId}.jpg`}
          alt={hero.superhero} 
          className="img-thumbnail "
        />
      </div>
      <div className="col-8">
        <h3>{hero.superhero}</h3>
        <ul className="list-group list-group-flush">
          <li className="list-group-item"><b>Alter ego:</b> {hero.alter_ego}</li>
          <li className="list-group-item"><b>Publisher</b> {hero.publisher}</li>
          <li className="list-group-item"><b>First Appearence</b> {hero.first_appearance}</li>
        </ul>
        <h5 className="mt-3">Characters</h5>
        <p>{hero.characters}</p>
        <button 
          onClick={onNavigateback}
          className="btn btn-outline-primary">
          Back
        </button>
      </div>
    </div>

  )
}
