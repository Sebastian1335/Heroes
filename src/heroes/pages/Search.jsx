import { useLocation, useNavigate } from "react-router-dom";
import { useForm } from "../../hooks/useForm";
import { HeroCard } from "../components";
import queryString from "query-string";
import { getHeroByName } from "../helpers/geHeroesByName";

export const Search = () => {
    const navigate = useNavigate();

    const location = useLocation(); //! Para leer querys dentro del URL

    const { q = "" } = queryString.parse(location.search); //! extrae los querys parameters
    const { searchText, onInputChange, onResetForm } = useForm({
        searchText: q,
    });

    const heroes = getHeroByName(q);

    const showSearch = (q.length === 0) //!  devuelve un booleano
    const showError = (q.length > 0) && heroes.length === 0;
    const onSearchSubmit = (event) => {
        event.preventDefault();
        navigate(`?q=${searchText}`); //!  Crea el query parameter
    };

    return (
        <>
            <h1>Search</h1>
            <hr />
            <div className="row">
                <div className="col-5">
                    <h4>
                        Searching
                        <hr />
                        <form onSubmit={onSearchSubmit} aria-label="form">
                            <input
                                type="text"
                                placeholder="Search a Hero"
                                className="form-control"
                                name="searchText"
                                autoComplete="off"
                                value={searchText}
                                onChange={onInputChange}
                            />
                            <button className="btn btn-outline-primary mt-1">
                                Search
                            </button>
                        </form>
                    </h4>
                </div>
                <div className="col-7">
                    <h4>Restuls</h4>
                    <hr />
                    {/* {q === "" ? (
                        <div className="alert alert-primary">Search a hero</div>
                    ) : (
                        heroes.length === 0 && (
                            <div className="alert alert-danger">
                                No hero with <b>{q}</b>
                            </div>
                        )
                    )} */}
                    <div aria-label="SearchDiv" className="alert alert-primary animate__animated animate__fadeIn" style={{display: showSearch ? '' : "none"}}>Search a hero</div>
                    <div aria-label="UnfoundDiv" className="alert alert-danger animate__animated animate__fadeIn" style={{display: showError ? '' : "none"}}>
                        No hero with <b>{q}</b>
                    </div>
                    {heroes.map((hero) => (
                        <HeroCard key={hero.id} {...hero} />
                    ))}
                </div>
            </div>
        </>
    );
};
