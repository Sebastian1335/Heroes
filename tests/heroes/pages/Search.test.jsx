import { fireEvent, render, screen } from "@testing-library/react"
import { MemoryRouter } from "react-router-dom"
import { Search } from "../../../src/heroes/pages/Search"


const MockedUsednavigate = jest.fn() //! Necesariamente debe tener el nombre Mocked...

jest.mock('react-router-dom',() => ({
    ...jest.requireActual('react-router-dom'),
    useNavigate: () => MockedUsednavigate
}))
describe('Pruebas en el <Search/>', () => {
    beforeEach(() => jest.clearAllMocks())
    test('debe de mostrarse correctamente con valores por defecto', () => {
        const {container} = render(
            <MemoryRouter>
                <Search/>
            </MemoryRouter>
        )
        expect(container).toMatchSnapshot()
    })
    test('Debe de mostrar a Batman y el input con el valor del queryString', () => {
        render(
            <MemoryRouter initialEntries={['/search?q=batman']}>
                <Search/>
            </MemoryRouter>
        )
        const input = screen.getByRole('textbox')
        expect(input.value).toBe('batman')
        const img = screen.getByRole('img')
        expect(img.src).toContain('/assets/heroes/dc-batman.jpg')
        const divSearch = screen.getByLabelText('SearchDiv')
        expect(divSearch.style._values).toEqual({"display": "none"})
    })
    test('Debe de mostrar un error si no se encuentra el hero (batman123)', () => {
        render(
            <MemoryRouter initialEntries={['/search?q=WASAA']}>
                <Search/>
            </MemoryRouter>
        )
        const input = screen.getByRole('textbox')
        expect(input.value).toBe('WASAA')
        const divUnfound = screen.getByLabelText('UnfoundDiv')
        expect(divUnfound.style.display).toBe("")
    })
    test('Debe de llamar el navigate a la pantalla nueva', () => {
        //* Disparar evento en la caja de texto para ingresar un valor
        //* Agarrar el formulario y disparar el  evento submit
        //* Espero que se mande a mandar el navigare con el argumento de superheroe
        render(
            <MemoryRouter initialEntries={['/search']}>
                <Search/>
            </MemoryRouter>
        )
        const input = screen.getByRole('textbox')
        fireEvent.change(input, {target: { name: 'searchText', value: 'superman'}})
        screen.debug()
        const form = screen.getByRole('form') //! no confundir getBy con findBy
        fireEvent.submit(form)
        expect(MockedUsednavigate).toHaveBeenCalledWith('?q=superman')

    })
})

//! Leer parametros del url con memory router