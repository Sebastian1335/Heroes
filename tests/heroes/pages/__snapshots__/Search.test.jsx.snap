// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el <Search/> debe de mostrarse correctamente con valores por defecto 1`] = `
<div>
  <h1>
    Search
  </h1>
  <hr />
  <div
    class="row"
  >
    <div
      class="col-5"
    >
      <h4>
        Searching
        <hr />
        <form
          aria-label="form"
        >
          <input
            autocomplete="off"
            class="form-control"
            name="searchText"
            placeholder="Search a Hero"
            type="text"
            value=""
          />
          <button
            class="btn btn-outline-primary mt-1"
          >
            Search
          </button>
        </form>
      </h4>
    </div>
    <div
      class="col-7"
    >
      <h4>
        Restuls
      </h4>
      <hr />
      <div
        aria-label="SearchDiv"
        class="alert alert-primary animate__animated animate__fadeIn"
      >
        Search a hero
      </div>
      <div
        aria-label="UnfoundDiv"
        class="alert alert-danger animate__animated animate__fadeIn"
        style="display: none;"
      >
        No hero with 
        <b />
      </div>
    </div>
  </div>
</div>
`;
